src/message/dkim.rs

```diff
@@ -0,0 +1,318 @@
+                    HeaderName::new_from_ascii(h).unwrap(),
+                    dkim_canonicalize_header_value(value, canonicalization),
+                ),
+                DkimCanonicalizationType::Relaxed => signed_headers_relaxed.push_str(&format!(
+                    "{}:{}",
+                    h,
+                    dkim_canonicalize_header_value(value, canonicalization)
+                )),
+            }
+        }
+    }
+    match canonicalization {
+        DkimCanonicalizationType::Simple => format!("{}", signed_headers),
+        DkimCanonicalizationType::Relaxed => signed_headers_relaxed,
+    }
+}
+
+/// Sign with Dkim a message by adding Dkim-Signture header created with configuration expressed by
+/// dkim_config
+///
+
+pub fn dkim_sign(message: &mut Message, dkim_config: DkimConfig) {
+    let timestamp = format!(
+        "{}",
+        SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+    );
+    let headers = message.headers();
+    let body_hash = Sha256::digest(
+        dkim_canonicalize_body(&message.body_raw(), dkim_config.canonicalization.body).as_bytes(),
+    );
+    let bh = encode(body_hash);
+    let signed_headers_list = match dkim_config.canonicalization.header {
+        DkimCanonicalizationType::Simple => dkim_config.headers.join(":"),
+        DkimCanonicalizationType::Relaxed => dkim_config.headers.join(":").to_lowercase(),
+    };
+    let dkim_header = dkim_header_format(
+        dkim_config.domain.clone(),
+        dkim_config.selector.clone(),
+        dkim_config.canonicalization,
+        timestamp.clone(),
+        signed_headers_list.clone(),
+        bh.clone(),
+        "".to_string(),
+    );
+    let signed_headers = dkim_canonicalize_headers(
+        dkim_config.headers.clone(),
+        headers,
+        dkim_config.canonicalization.header,
+    );
+    let private_key = RsaPrivateKey::from_pkcs1_pem(&dkim_config.private_key).unwrap();
+    let canonicalized_dkim_header = dkim_canonicalize_headers(
+        vec!["DKIM-Signature".to_string()],
+        &dkim_header,
+        dkim_config.canonicalization.header,
+    );
+    let canonicalized_dkim_header = canonicalized_dkim_header.trim_end();
+    let to_be_signed = format!("{}{}", signed_headers, canonicalized_dkim_header);
+    let to_be_signed = to_be_signed.trim_end();
+    let hashed_headers = Sha256::digest(to_be_signed.as_bytes());
+    let signature = encode(
+        private_key
+            .sign(
+                PaddingScheme::new_pkcs1v15_sign(Some(Hash::SHA2_256)),
+                &hashed_headers,
+            )
+            .unwrap(),
+    );
+    let dkim_header = dkim_header_format(
+        dkim_config.domain,
+        dkim_config.selector,
+        dkim_config.canonicalization,
+        timestamp,
+        signed_headers_list,
+        bh,
+        signature,
+    );
+    let mut headers = headers.clone();
```

Comment:
Why do we clone `message.headers` just to write it back to `message.headers`? Can we mutate that directly?